server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregatorTests.java

```diff
@@ -338,6 +357,18 @@ public void testSelfReferencingAggStateAfterCombine() throws IOException {
         }
     }
 
+    @Override
+    @SuppressWarnings("unchecked")
+    protected <A extends Aggregator> A createAggregator(Query query,
+                                                        AggregationBuilder aggregationBuilder,
+                                                        IndexSearcher indexSearcher,
+                                                        IndexSettings indexSettings,
+                                                        MultiBucketConsumerService.MultiBucketConsumer bucketConsumer,
+                                                        MappedFieldType... fieldTypes) throws IOException {
+        aggregator = spy(super.createAggregator(query, aggregationBuilder, indexSearcher, indexSettings, bucketConsumer, fieldTypes));
```

Comment:
Im not particularly a fan of using `spy()` here. There are a couple of reasons for this:
1. The class under is being wrapped by `Mockito` which feels a bit weird to me because we are not really testing the class we intend to test. It also has the danger of the `Mockito` usage being extended in future and us accidentally bypassing the actual logic in the `ScriptedMetricAggregator` itself
2. We are creating a `ensureNoSelfReferencesInAggState()` method purely so we can do this spying which feels a bit ugly to me.

I wonder if we need to do this spying at all and instead could maybe just rely on the existing tests that check we catch when a self-reference occurs and not worry about ensuring its only called once? It feels to me like the downsides of doing this mockito check might outweigh the benefits. 

@rjernst do you have any thoughts on this?

Reply:
I understand and I do agree with you @colings86.

Another idea I can suggest would be to throw an exception from the `ScriptedMetricAggregator` if we do ensure no self references in agg state more than once.

I let @rjernst give us his opinion on this.

Reply:
I agree with @colings86. I would rely on/improve existing tests. Mocking should not be necessary here. A mock script can be added to the existing tests which creates a cycle, and ScriptedMetricAggregator can call the normal ensureNoSelfReferences method directly.

Reply:
Thank you @rjernst, I'll remove unnecessary mocking.